### ISSUE TITLE
C++11 user defined string literals break the json file

### ISSUE THREAD

**Issue by kobi-ca** [namhyung/uftrace#1129]
and it cannot be loaded.
C++11 user defined string literals looks like the following:
```
ret operator "" _name(some_type var) {
// ...
}

// to make it work with
auto v = 10_name;
```
in the json, the name of the func would be
```
"operator ""'
```
those 3 trailing quotes makes it impossible to load this into chrome

**Comment by honggyukim** [comment]
@kobi-ca Thanks for the report, but could you please provide us a simple and complete example program that shows the problem?

**Comment by kobi-ca** [comment]
sure. will do. And I also need to get back to you on another issue that you resolved. sorry for the delay. Will make it happen this week. take care!

**Comment by kobi-ca** [comment]
Here is an example:
```cpp
#include <iostream>

namespace my_user_literals {

long double operator"" _kg(long double x) { return x * 1000; }

} // namespace my_user_literals

// Driver code
int main() {
  using namespace my_user_literals;
  long double weight = 3.6_kg;
  std::cout << weight << std::endl;
  return 0;
}

```

cmake:
```cmake
cmake_minimum_required(VERSION 3.15)
project(cpp_user_literals_example)

set(CMAKE_CXX_STANDARD 14)

add_executable(cpp_user_literals_example main.cpp)

target_compile_options(cpp_user_literals_example PRIVATE -finstrument-functions -fno-omit-frame-pointer -ggdb)
```

running
```
uftrace record ./cpp_user_literals_example
```

dumping
```
uftrace dump --chrome
{"traceEvents":[
{"ts":0,"ph":"M","pid":32733,"name":"process_name","args":{"name":"[32733] cpp_user_litera"}},
{"ts":0,"ph":"M","pid":32733,"name":"thread_name","args":{"name":"[32733] cpp_user_litera"}},
{"ts":1124421380161.629,"ph":"B","pid":32733,"name":"_GLOBAL__sub_I_my_user_literals::operator"""},
{"ts":1124421380161.963,"ph":"B","pid":32733,"name":"__static_initialization_and_destruction_0"},
{"ts":1124421380162.589,"ph":"B","pid":32733,"name":"std::ios_base::Init::Init"},
{"ts":1124421380284.084,"ph":"E","pid":32733,"name":"std::ios_base::Init::Init"},
{"ts":1124421380285.454,"ph":"B","pid":32733,"name":"__cxa_atexit"},
{"ts":1124421380286.132,"ph":"E","pid":32733,"name":"__cxa_atexit"},
{"ts":1124421380286.428,"ph":"E","pid":32733,"name":"__static_initialization_and_destruction_0"},
{"ts":1124421380286.682,"ph":"E","pid":32733,"name":"_GLOBAL__sub_I_my_user_literals::operator"""},
{"ts":1124421380287.189,"ph":"B","pid":32733,"name":"main"},
{"ts":1124421380290.063,"ph":"B","pid":32733,"name":"my_user_literals::operator"""},
{"ts":1124421380290.144,"ph":"E","pid":32733,"name":"my_user_literals::operator"""},
{"ts":1124421380290.616,"ph":"B","pid":32733,"name":"std::basic_ostream::operator<<"},
{"ts":1124421380325.236,"ph":"E","pid":32733,"name":"std::basic_ostream::operator<<"},
{"ts":1124421380325.538,"ph":"B","pid":32733,"name":"std::basic_ostream::operator<<"},
{"ts":1124421380343.679,"ph":"E","pid":32733,"name":"std::basic_ostream::operator<<"},
{"ts":1124421380343.897,"ph":"E","pid":32733,"name":"main"}
], "displayTimeUnit": "ns", "metadata": {
"version":"uftrace v0.9 ( dwarf python tui perf sched )",
"recorded_time":"Thu Feb 20 16:08:36 2020",
"command_line":"uftrace record ./cpp_user_literals_example"
} }
```

note the line 
`{"ts":1124421380161.629,"ph":"B","pid":32733,"name":"_GLOBAL__sub_I_my_user_literals::operator"""},`
and similar lines.
if you try to load it into `chrome://tracing` it would fail since this is not a valid json.

If I escape it (i.e. having \ before the internal double quotes, then it works). e.g.
`{"ts":1124421380161.629,"ph":"B","pid":32733,"name":"_GLOBAL__sub_I_my_user_literals::operator\"\""},`

Please let me know if you have any question,
thanks,
Kobi

**Comment by honggyukim** [comment]
Hi @kobi-ca,

Thanks for providing the reproducible example. I've found the problem and the following change will fix the problem.

You can also download the diff here - [json_escaped_name.diff.txt](https://github.com/namhyung/uftrace/files/4233816/json_escaped_name.diff.txt)

```diff
diff --git a/cmds/dump.c b/cmds/dump.c
index 01123801..dbd2d41b 100644
--- a/cmds/dump.c
+++ b/cmds/dump.c
@@ -855,6 +855,20 @@ static void dump_chrome_header(struct uftrace_dump_ops *ops,
        chrome->last_comma = false;
 }

+static inline void fill_json_escaped_name(char *json_name, const char *name)
+{
+       int len, i, j;
+
+       j = 0;
+       len = strlen(name);
+       for (i = 0; i < len; i++) {
+               if (name[i] == '"')
+                       json_name[j++] = '\\';
+               json_name[j++] = name[i];
+       }
+       json_name[j] = '\0';
+}
+
 static void dump_chrome_task_rstack(struct uftrace_dump_ops *ops,
                                     struct uftrace_task_reader *task, char *name)
 {
@@ -865,6 +879,7 @@ static void dump_chrome_task_rstack(struct uftrace_dump_ops *ops,
        struct uftrace_chrome_dump *chrome = container_of(ops, typeof(*chrome), ops);
        bool is_process = task->t->pid == task->tid;
        int rec_type = frs->type;
+       char json_name[1024];

        if (rec_type == UFTRACE_EVENT) {
                switch (frs->addr) {
@@ -889,16 +904,18 @@ static void dump_chrome_task_rstack(struct uftrace_dump_ops *ops,
                pr_out(",\n");
        chrome->last_comma = true;

+       fill_json_escaped_name(json_name, name);
+
        if (rec_type == UFTRACE_ENTRY) {
                ph = 'B';
                if (is_process) {
                        /* no need to add "tid" field */
                        pr_out("{\"ts\":%"PRIu64".%03d,\"ph\":\"%c\",\"pid\":%d,\"name\":\"%s\"",
-                              frs->time / 1000, (int)(frs->time % 1000), ph, task->tid, name);
+                              frs->time / 1000, (int)(frs->time % 1000), ph, task->tid, json_name);
                }
                else {
                        pr_out("{\"ts\":%"PRIu64".%03d,\"ph\":\"%c\",\"pid\":%d,\"tid\":%d,\"name\":\"%s\"",
-                              frs->time / 1000, (int)(frs->time % 1000), ph, task->t->pid, task->tid, name);
+                              frs->time / 1000, (int)(frs->time % 1000), ph, task->t->pid, task->tid, json_name);
                }

                if (frs->more) {
@@ -914,11 +931,11 @@ static void dump_chrome_task_rstack(struct uftrace_dump_ops *ops,
                if (is_process) {
                        /* no need to add "tid" field */
                        pr_out("{\"ts\":%"PRIu64".%03d,\"ph\":\"%c\",\"pid\":%d,\"name\":\"%s\"",
-                              frs->time / 1000, (int)(frs->time % 1000), ph, task->tid, name);
+                              frs->time / 1000, (int)(frs->time % 1000), ph, task->tid, json_name);
                }
                else {
                        pr_out("{\"ts\":%"PRIu64".%03d,\"ph\":\"%c\",\"pid\":%d,\"tid\":%d,\"name\":\"%s\"",
-                              frs->time / 1000, (int)(frs->time % 1000), ph, task->t->pid, task->tid, name);
+                              frs->time / 1000, (int)(frs->time % 1000), ph, task->t->pid, task->tid, json_name);
                }

                if (frs->more) {
```
However, we might have to handle more cases so need to consider if there are other candidate special characters.

It only handles `"` character, but we need to care other operators that can be overloaded. Could you please provide the list of other possible candidate operators except for `operator""`?

**Comment by honggyukim** [comment]
It seems that other operators do not have such problems. It's because of json parsing specific problem that contains `"` while it understands the name boundary with the same character.

The above change fixes the problem as follows:

**Before**
```
    ...
{"ts":1810288566826.266,"ph":"B","pid":18337,"name":"my_user_literals::operator"""},
{"ts":1810288566826.316,"ph":"E","pid":18337,"name":"my_user_literals::operator"""},
    ...
```
**After**
```
    ...
{"ts":1810288566826.266,"ph":"B","pid":18337,"name":"my_user_literals::operator\"\""},
{"ts":1810288566826.316,"ph":"E","pid":18337,"name":"my_user_literals::operator\"\""},
    ...
```

**Comment by kobi-ca** [comment]
thanks for fixing it. look good.
Probably need to look at this:
https://www.freeformatter.com/json-escape.html

for now, this would be enough, I think:
```
Double quote is replaced with \"
Backslash is replaced with \\
```
you can regex it in your code to achieve it. (but you are .c and not .cpp so not sure if you have access to some standard regex lib in C)

**Comment by honggyukim** [comment]
Hmm.. Backslash cannot be used as a function name, so I think handling double quote is enough. If @namhyung agrees, I will send a PR. @kobi-ca, thanks for your help.

**Comment by namhyung** [comment]
We have `print_json_escaped_char()` in cmd/replay.c - any chance to reuse the code?

**Comment by honggyukim** [comment]
This problem was already fixed by 6cc2bedfd35a2c7e6db289cd1e3120ef2804c172, but #1230 refactors to reuse `print_json_escaped_char` as @namhyung suggested.
